src/main/kotlin/uk/gov/dluhc/printapi/config/CorrelationIdMdcConfiguration.kt

```diff
@@ -0,0 +1,110 @@
+        ex: Exception?
+    ) {
+        MDC.remove(CORRELATION_ID)
+    }
+}
+
+/**
+ * AOP Aspect to read and set the correlation ID on inbound (received) and outbound SQS [Message]s respectively.
+ * This allows for passing and logging a consistent correlation ID between disparate systems or processes.
+ */
+@Aspect
+@Component
+class CorrelationIdMdcMessageListenerAspect {
+
+    /**
+     * Before Advice for inbound [Message]s (ie. SQS Message's being directed to a listener class) that sets the correlation ID
+     * MDC variable to the value found in the Message header `x-correlation-id` if set, or a new value.
+     * This allows for passing and logging a consistent correlation ID between disparate systems or processes.
+     */
+    @Before("execution(* org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler.handleMessage(..))")
+    fun beforeHandleMessage(joinPoint: JoinPoint) {
+        val message = joinPoint.args[0] as Message<*>?
+        MDC.put(CORRELATION_ID, message?.headers?.get(CORRELATION_ID_HEADER)?.toString() ?: generateCorrelationId())
+    }
+
+    /**
+     * Around Advice for outbound [Message]s (ie. SQS Message's being sent) that sets the correlation ID
+     * header on a new [Message] to either the existing MDC variable or a new value if not set in MDC.
+     * This allows for passing and logging a consistent correlation ID between disparate systems or processes.
+     *
+     * The reason this Advice is an Around is because [Message] and it's headers are immutable, so we cannot add
+     * the correlation ID header on the passed [Message]. Therefore we need to create new message with the same
+     * payload and a modified collection of headers.
+     */
+    @Around("execution(* io.awspring.cloud.messaging.core.support.AbstractMessageChannelMessagingSendingTemplate.send(..))")
+    fun aroundSendMessage(proceedingJoinPoint: ProceedingJoinPoint): Any? {
+        val queue = proceedingJoinPoint.args[0]
+        val originalMessage = proceedingJoinPoint.args[1] as Message<*>
+        val newMessage = GenericMessage(
+            originalMessage.payload,
+            originalMessage.headers.toMutableMap().plus(CORRELATION_ID_HEADER to getCurrentCorrelationId())
+        )
+        return proceedingJoinPoint.proceed(arrayOf(queue, newMessage))
+    }
+}
+
+/**
+ * AOP Aspect for Scheduled tasks (ie. cron tasks) that sets the correlation ID MDC variable.
+ */
+@Aspect
+@Component
+class CorrelationIdMdcScheduledAspect {
+
+    /**
+     * Before Advice for Scheduled tasks (ie. cron tasks) that sets the correlation ID MDC variable to a new value.
+     * Due to the invocation semantics of a Scheduled task it does not make sense to pass a correlation ID from another
+     * system or process into it.
+     */
+    @Before("@annotation(org.springframework.scheduling.annotation.Scheduled)")
+    fun before(joinPoint: JoinPoint) {
+        MDC.put(CORRELATION_ID, generateCorrelationId())
```

Comment:
Where is this removed?

Reply:
Hmmm, they're not currently .... I was thinking this might be OK as I was assuming each cron invocation and SQS handling (the other one you commented on) were on a thread in their own right, and that MDC was thread local ... and that it might be OK?
But perhaps not? 🤔 

Do you think these `@Before` advices might be better off being `@Around`, and then we set the MDC, invoke the method, then clear the MDC ?
Would that be a better approach ?

Reply:
They are thread local, but if that thread is reused and the correlation id is not reset it can become misleading.  Always best to unset the thread local once done with it.  I think your `@Around` suggestion makes sense.

Reply:
OK, I've changed these advices to `@Around` in order to clear the MDC 👍